### DISCUSSION TITLE
Seg Fault when running RapidWright in WSL

### DISCUSSION THREAD

**Discussion by ghostlyinfluence** [Xilinx/RapidWright#552] [Q&A] [answered]
Hello RW community,

I am currently trying to use RapidWright in Ubuntu 20.04 LTS that is installed into a WSL 2 (Windows Sub-System for Linux) instance. The other RW tools have worked great but I am now trying to go through the PicoBlaze tutorial which requires usage of the RW performance explorer and Vivado Seg Faults every time with the following error:

***************************************************************************
* ERROR: Job 48 failed
* LOG FILE: /mnt/d/Development/rwdev/picoblaze/Explore_Explore_-0.05_pblock0/cmd.log
*  Here are the last few lines of the log:
Finished Parsing XDC File [/mnt/d/Development/rwdev/picoblaze/initial/picoblaze_top_late.xdc]
Reading XDEF placement.
Reading placer database...
Reading XDEF routing.
Read XDEF File: Time (s): cpu = 00:00:00.26 ; elapsed = 00:00:07 . Memory (MB): peak = 2659.250 ; gain = 0.000 ; free physical = 392 ; free virtual = 1891
Restored from archive | CPU: 0.450000 secs | Memory: 0.247276 MB |
Finished XDEF File Restore: Time (s): cpu = 00:00:00.26 ; elapsed = 00:00:07 . Memory (MB): peak = 2659.250 ; gain = 0.000 ; free physical = 327 ; free virtual = 1835
/tools/Xilinx/Vivado/2021.2/bin/rdiArgs.sh: line 311: 22832 Killed                  "$RDI_PROG" "$@"
***************************************************************************
Failed Job Count: 60
Performance Explorer Failed!

I guess I am willing to accept that this is just user error and the performance explorer cannot be run in a VM or VM like environment but I figured it was worth discussion before trying to move to a native Linux box.

**Things I have tried:**
1. Launching with one CPU core to try and minimize memory utilization using the -z 1 flag | Didn't help
2. Launching with a custom Vivado path to increase stack size and try to avoid a memory induced seg fault -y "vivado -stack 2000" | Didn't help
3. Tried fixing a potentially broken library by following [this guide](https://support.xilinx.com/s/question/0D52E00006hpSCpSAM/vitis-on-wsl?language=en_US) | Didn't help
4. Tried cutting down the amount of fabric to explore from 180 -> 120 -> 60 jobs | Didn't help

Other system info:
CPU Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
RAM: 64GB

Thanks for the help and discussion!

**Comment by clavin-xlnx** [comment, marked as answer]
RapidWright and Vivado are supported on Windows, is there a particular constraint in your workflow that requires WSL?  How did you install Vivado for WSL?  

Although we might be able to accommodate minor changes for RapidWright on WSL, it would be harder to make changes to Vivado for that platform if something wasn't working correctly.   Your best bet might be to try it on a native Linux or Windows machine.

**Reply by ghostlyinfluence** [reply]
Chris,

Thank you for the quick response. No I guess there isn't a reason that I needed to be running Vivado and RapidWright in WSL vs Windows. I didn't have an additional machine available to run Linux natively which is why I went to WSL. I guess I've always done dev in Linux and figured it was required for RW and Vivado.

My assumption is that this isn't directly an issue with RapidWright and more of an issue with Vivado. Although maybe the way that you call Vivado instances in the Performance Explorer results in memory issues with Vivado in virtualized environments.

I'll take a look at running in Windows or see if I can't find another machine to run native Linux and see how that goes.

Thanks,

**Reply by ghostlyinfluence** [reply]
Hey Calvin,

Well it seems like switching to a native Linux OS installation did not help as I still seg-faulted although the actual readout on the command line was slightly different.

"
Waiting on 40 running, 0 not yet started...
Waiting on 39 running, 0 not yet started...
Waiting on 37 running, 0 not yet started...
Waiting on 34 running, 0 not yet started...
Waiting on 21 running, 0 not yet started...
Waiting on 0 not yet started...
***************************************************************************
* ERROR: Job 26 failed
* LOG FILE: /home/canyonlands/Development/picoblaze/Default_Explore_0.125_pblock0/cmd.log
*  Here are the last few lines of the log:
INFO: [Device 21-403] Loading part xczu3eg-sbva484-1-e
Netlist sorting complete. Time (s): cpu = 00:00:00 ; elapsed = 00:00:00.03 . Memory (MB): peak = 2606.125 ; gain = 0.000 ; free physical = 3459 ; free virtual = 5889
INFO: [Netlist 29-17] Analyzing 63 Unisim elements for replacement
INFO: [Netlist 29-28] Unisim Transformation completed in 0 CPU seconds
INFO: [Project 1-479] Netlist was created with RapidWright 2022.1.3
INFO: [Project 1-570] Preparing netlist for logic optimization
Parsing XDC File [/home/canyonlands/Development/picoblaze/initial/picoblaze_top.xdc]
/tools/Xilinx/Vivado/2021.2/bin/rdiArgs.sh: line 311:  4909 Killed                  "$RDI_PROG" "$@"
***************************************************************************
Failed Job Count: 60
Performance Explorer Failed!
"

Do you have any other suggestions to try out? I am sort of at a loss here for what to try next as this seems to be a seg-fault within Vivado and not specifically RapidWright.

Colin

**Reply by clavin-xlnx** [reply]
Sorry for trouble you are seeing.  Here are a few things you could try to diagnose.  

1. From the message, it is clear at least one job failed, but its possible that many other jobs succeeded.  It might be worth checking the logs of some of the other runs to get an idea if this is complete failure or just a few runs.  Won't solve the problem, but you might be able to keeping going.
2. It looks like you are running 2021.2, perhaps try again with a different version such as 2022.1.  
3. By default, the `PerformanceExplorer` will run 4 threads in parallel on the host machine.  It is possible too many are running and then somehow causing an issue (overrun of memory, etc).  Perhaps trying running the `PerformanceExplorer` with just a couple jobs and one variation:
```
java com.xilinx.rapidwright.util.PerformanceExplorer -c clk -i picoblaze_synth.dcp -t 2.85 -b picoblaze_pblocks.txt -z 1 -p Default -r Default -u 0.0
```
This should only run 3 jobs in series and you can even go into the pblock file and remove two of them to get down to one.  If you can get a successful run, try building up until it breaks.  

If you are still having trouble, can you please provide the version of Linux you are using and as much information as possible about your environment so we can try to reproduce on our side?

**Reply by ghostlyinfluence** [reply]
Hi Calvin,

So initially, I think I found a user error... The performance explorer crashes out if you already have a pblock saved to your DCP file that matches any of the default names the explorer wants to use [pblock0, pblock1, etc.]. After deleting any pblocks that I had saved to my design I re-ran the modified script you provided <java com.xilinx.rapidwright.util.PerformanceExplorer -c clk -i picoblaze_synth.dcp -t 2.85 -b picoblaze_pblocks.txt -z 1 -p Default -r Default -u 0.0> and **SUCCESS**

I moved up to 4 total jobs <java com.xilinx.rapidwright.util.PerformanceExplorer -c clk -i picoblaze_synth.dcp -t 2.85 -b picoblaze_pblocks.txt -z 1 -u 0.0> **SUCCESS**

Then 15 total jobs <java com.xilinx.rapidwright.util.PerformanceExplorer -c clk -i picoblaze_synth.dcp -t 2.85 -b picoblaze_pblocks.txt -z 1 -p Default -r Default> **SUCCESS**

Then 30 total jobs <java com.xilinx.rapidwright.util.PerformanceExplorer -c clk -i picoblaze_synth.dcp -t 2.85 -b picoblaze_pblocks.txt -z 1 -p Default> **SUCCESS**

And then 60 total jobs <java com.xilinx.rapidwright.util.PerformanceExplorer -c clk -i picoblaze_synth.dcp -t 2.85 -b picoblaze_pblocks.txt -z 1> **FAILED**
This job failed in the same way as before.
Waiting on 2 running, 0 not yet started...
Waiting on 0 not yet started...
***************************************************************************
* ERROR: Job 8 failed
* LOG FILE: /home/canyonlands/Development/picoblaze/Default_Default_0.075_pblock0/cmd.log
*  Here are the last few lines of the log:
Phase 1.1 Placer Initialization Netlist Sorting | Checksum: 0fa2f23c

Time (s): cpu = 00:00:00.01 ; elapsed = 00:00:00.40 . Memory (MB): peak = 3495.328 ; gain = 0.000 ; free physical = 412 ; free virtual = 4386
INFO: [Opt 31-138] Pushed 0 inverter(s) to 0 load pin(s).
Netlist sorting complete. Time (s): cpu = 00:00:00 ; elapsed = 00:00:00 . Memory (MB): peak = 3495.328 ; gain = 0.000 ; free physical = 405 ; free virtual = 4377

Phase 1.2 IO Placement/ Clock Placement/ Build Placer Device
/tools/Xilinx/Vivado/2021.2/bin/rdiArgs.sh: line 311: 1053398 Killed                  "$RDI_PROG" "$@"
***************************************************************************

You were right about some jobs succeeding. I didn't click through all 60 but my feeling is that the majority of the jobs succeeded. One interesting thing to note is that even when passing the -z 1 flag, to limit the max number of concurrent jobs, I feel like it is being ignored. For example, Vivado still uses the max about of CPUs (in this case 8) and a thread for each RW Performance Explorer job gets created and is being executed concurrently; usually ~20 system tasks get created and ~400 new threads are created. I can verify the Vivado CPU usage in the cmd.log files that are created for each run and I look at HTOP to see the total amount of new threads and tasks.

![image](https://user-images.githubusercontent.com/53321206/194977255-fb8fb8f6-e51b-4ecd-a587-827e7b0b499c.png)

_Ultimately, this is still workable as I can limit the total jobs per run of the explorer to less than 30 and just iterate through some of the options on my own but I would be interested to understand if the -z flag is working as intended or if there was a way to limit Vivado CPU count from this application._

I am happy to provide more information if you think there is something here worth digging into.

Thanks,
Colin

**Reply by clavin-xlnx** [reply]
Hi Colin,

You were right in that `PerformanceExplorer` was running more jobs than was specified.  I've created a PR in #561 to address this issue.  

Thanks for bringing it up!

Chris
